### PR TITLE
[Train] [release test] Release tests for ray train local mode

### PR DESCRIPTION
**Pull request by xinyuangui2**
## Why are these changes needed?

Release test for ray train local mode.

Setup:

- 4 nodes, each with 4 gpus.
- Each node runs `torchrun`.

Test job: https://buildkite.com/ray-project/release/builds/61135

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

---

> [!NOTE]
> Adds a nightly Train release test that launches torchrun across 2 GPU nodes to run a Ray Train TorchTrainer on FashionMNIST and validates loss.
> 
> - **Train tests**:
>   - Add `torch_local_mode` to `release/release_tests.yaml` (nightly, BYOD GPU) running `python torch_local_mode_launcher.py` with cluster config `train_tests/local_mode/compute_gpu_2x4_aws.yaml`.
> - **New local mode assets**:
>   - `release/train_tests/local_mode/compute_gpu_2x4_aws.yaml`: 1 head `m5.4xlarge`, 2 workers `g4dn.12xlarge` (non-spot) in `us-west-2`.
>   - `release/train_tests/local_mode/torch_local_mode_launcher.py`: Starts a Ray cluster task that uploads `torch_local_mode_test.py` to 2 nodes and launches `torchrun` (`--nnodes=2`, `--nproc-per-node=4`) with `RAY_TRAIN_V2_ENABLED=1`.
>   - `release/train_tests/local_mode/torch_local_mode_test.py`: Defines a `TorchTrainer` training `resnet18` on FashionMNIST, reports checkpoints/metrics over 20 epochs, and asserts loss decreases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 11570687555ff507569859f7a6feea810cc921f0. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: release/train_tests/local_mode/torch_local_mode_launcher.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+
+@ray.remote
+def _torch_run_launch(
+    master_address: str,
+    node_rank: int,
+    absolute_path: str,
+    n_nodes: int = 4,
+    n_processes_per_node: int = 4,
+    master_port: str = 29500,
```
Comment (JasonLi1909) [inline release/train_tests/local_mode/torch_local_mode_launcher.py]:
```suggestion
    master_port: int = 29500,
```

File: release/train_tests/local_mode/torch_local_mode_launcher.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+):
+    cmd_node1 = [
```
Comment (JasonLi1909) [inline release/train_tests/local_mode/torch_local_mode_launcher.py]: nit: rename to `cmd`

File: release/train_tests/local_mode/compute_gpu_4x4_aws.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+head_node_type:
+    name: head_node
+    instance_type: m5.4xlarge
+
+worker_node_types:
+    - name: worker_node
+      instance_type: g4dn.12xlarge
+      max_workers: 4
+      min_workers: 4
```
Comment (JasonLi1909) [inline release/train_tests/local_mode/compute_gpu_4x4_aws.yaml]: Is there a reason we need 4 nodes, why not just 2?

**Review by JasonLi1909 — APPROVED** [review]

File: release/train_tests/local_mode/compute_gpu_4x4_aws.yaml (hunk 3 above)
Comment (xinyuangui2) [reply release/train_tests/local_mode/compute_gpu_4x4_aws.yaml]: Changed to 2

File: release/train_tests/local_mode/torch_local_mode_test.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
+"""Ray train release test: local mode launched by torchrun
+"""
```
Comment (matthewdeng) [inline release/train_tests/local_mode/torch_local_mode_test.py]: Can you add more details about what are the main functionalities that this test is covering?

File: release/train_tests/local_mode/torch_local_mode_test.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    # Train the model.
+    result = trainer.fit()
+
+    # Inspect the results.
+    final_loss = result.metrics["loss"]
+    logger.info(f"final_loss: {final_loss}")
+
+
+if __name__ == "__main__":
+    fit_func()
```
Comment (matthewdeng) [inline release/train_tests/local_mode/torch_local_mode_test.py:162]: Is this run anywhere?

Would it be helpful to verify that the runtimes are the same? And maybe set the seeds and compare the results?

File: release/train_tests/local_mode/torch_local_mode_launcher.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        f.write(stream)
+
+
+@ray.remote
+def _torch_run_launch(
+    master_address: str,
+    node_rank: int,
+    absolute_path: str,
+    n_nodes: int = 2,
+    n_processes_per_node: int = 4,
```
Comment (matthewdeng) [inline release/train_tests/local_mode/torch_local_mode_launcher.py]: nit: I don't think default values make sense here.

File: release/train_tests/local_mode/torch_local_mode_test.py (hunk 5 above)
Comment (xinyuangui2) [reply release/train_tests/local_mode/torch_local_mode_test.py:162]: Added loss convergence check.

**Review by matthewdeng — APPROVED** [review]
